File: Python/backend_fastapi/shared/modules/user_module.py
from infra.db.inmemory.mappers.user_inmemory_mapper import UserInMemoryMapper
from infra.db.inmemory.repositories.user_inmemory_repository import UserInMemoryRepository
from infra.db.inmemory.data.user_db import user_db
from shared.dto.mappers.user_dto_mapper import UserDTOMapper
from services.user_service import UserService

_user_db_mapper = UserInMemoryMapper()
_user_repository = UserInMemoryRepository(database=user_db, mapper=_user_db_mapper)

_user_service = UserService(repository=_user_repository)
_user_dto_mapper = UserDTOMapper()

user_module = {
    "service": _user_service,
    "dto_mapper": _user_dto_mapper,
}

File: Javascript/backend_nest/src/framework/modules/UserModule.ts
import { Module } from '@nestjs/common';
import { UserController } from '../controllers/UserController';
import { UserDTOMapper } from 'src/dto/mapper/UserDTOMapper';
import { UserInMemoryMapper } from 'src/infra/db/inmemory/mappers/UserInMemoryMapper';
import { userInMemoryDB } from 'src/infra/db/inmemory/data/UserDB';
import { UserInMemoryRepository } from 'src/infra/db/inmemory/repositories/UserInMemoryRepository';
import { UserRepository } from 'src/domain/repositories/UserRepository';
import { UserService } from 'src/services/UserService';
import { Mapper } from 'src/domain/kernel/Mapper';
import { User } from 'src/domain/entities/User';

@Module({
  imports: [],
  controllers: [UserController],
  providers: [
    {
      provide: 'DTOMapper',
      useClass: UserDTOMapper, // On injecte le mapper pour les DTOs
    },
    {
      provide: 'DatabaseMapper',
      useClass: UserInMemoryMapper, // Le moment où on passe sur autre chose que du inmemory, on peut changer cette ligne pour l'injection
    },
    {
      provide: 'Repository',
      useFactory: (mapper: Mapper<User, any>) => new UserInMemoryRepository(userInMemoryDB, mapper), // Idem ici, on change juste la classe de repository
      inject: ['DatabaseMapper'],
    },
    {
      provide: 'Service',
      useFactory: (authorRepository: UserRepository) => new UserService(authorRepository),
      inject: ['Repository'],
    },
  ],
})
export class UserModule {}
